onnx/checker.cc

```diff
@@ -0,0 +1,247 @@
+#include "onnx/checker.h"
+
+#include "onnx/defs/schema.h"
+#include "onnx/utils.h"
+
+#include <unordered_set>
+
+namespace onnx {
+namespace checker {
+
+#define fail(...) throw ValidationError(onnx::MakeString(__VA_ARGS__));
+
+#define enforce_has_field(proto, field)                                    \
+  if (!proto.has_##field()) {                                              \
+    fail("Field '", #field, "' of ", #proto, " is required but missing."); \
+  }
+
+#define enforce_has_repeated_field(proto, field)                    \
+  if (!proto.field##_size()) {                                      \
+    fail("Repeated Field '", #field, "' is required but missing."); \
+  }
+
+#define enforce_non_empty_field(proto, field)                                  \
+  if (proto.field().empty()) {                                                 \
+    fail("Field '", #field, "' of ", #proto, " is required to be non-empty."); \
+  }
+
+void check_value_info(const ValueInfoProto& value_info, int ir_version) {
+  enforce_non_empty_field(value_info, name);
+  enforce_has_field(value_info, type);
+  const auto value_case = value_info.type().value_case();
+  switch (value_case) {
+    case TypeProto::kTensorType: {
+      const auto& type = value_info.type().tensor_type();
+      enforce_has_field(type, elem_type);
+      enforce_has_field(type, shape);
```

Comment:
we may relax the shape check? for internal nodes' outputs, the shape may be unknown at the beginning, and it may be inferred via a shape inference through the whole graph after loading a graph.

Reply:
If enforced to have a empty shape initialized, then the tensor itself may be recognized as a scalar later as it has shape with zero dimension.

Reply:
it's not called to check the internal nodes right now. but I think even we do in the future, it makes senses the ValueInfo of an internal node to be missing in the model pb file instead of having an incomplete ValueInfo